ctapipe_io_lst/calibration.py

```diff
@@ -500,123 +520,191 @@ def interpolate_spikes(self, event, tel_id):
         """
         run_id = event.lst.tel[tel_id].svc.configuration_id
 
-        # Interpolate spikes should be done after pedestal subtraction and time lapse correction.
-        if isinstance(event.r1.tel[tel_id].waveform, np.ndarray):
-            waveform = event.r1.tel[tel_id].waveform.copy()
+        r1 = event.r1.tel[tel_id]
+        if r1.selected_gain_channel is None:
+            interpolate_spikes(
+                waveform=r1.waveform,
+                first_capacitors=self.first_cap[tel_id],
+                previous_first_capacitors=self.first_cap_old[tel_id],
+                run_id=run_id,
+            )
+        else:
+            interpolate_spikes_gain_selected(
+                waveform=r1.waveform,
+                first_capacitors=self.first_cap[tel_id],
+                previous_first_capacitors=self.first_cap_old[tel_id],
+                selected_gain_channel=r1.selected_gain_channel,
+                run_id=run_id,
+            )
+
+
+def convert_to_pe(waveform, calibration, selected_gain_channel):
+    if selected_gain_channel is None:
+        waveform -= calibration.pedestal_per_sample[:, :, np.newaxis]
+        waveform *= calibration.dc_to_pe[:, :, np.newaxis]
+    else:
+        waveform -= calibration.pedestal_per_sample[selected_gain_channel, PIXEL_INDEX, np.newaxis]
+        waveform *= calibration.dc_to_pe[selected_gain_channel, PIXEL_INDEX, np.newaxis]
+
+@njit(cache=True)
+def interpolate_spike_A(waveform, position):
+    """
+    Numba function for interpolation spike type A.
+    Change waveform array.
+    """
+    a = int(waveform[position - 1])
+    b = int(waveform[position + 2])
+    waveform[position] = waveform[position - 1] + (0.33 * (b - a))
+    waveform[position + 1] = waveform[position - 1] + (0.66 * (b - a))
+
+
+@njit(cache=True)
+def interpolate_spikes_pixel(waveform, current_fc, last_fc):
```

Comment:
same thing here, you changed the name of the function and parameters, but body of the function does not seem to be much different, it makes the diff quite messy

Reply:
I broke this function out of the large function also containing the loops, because this is the same whether gain selection has already happend or not, only the loop is different. No real way to avoid the not-so-nice diff I think